src/opnsense/scripts/unbound/blocklists/__init__.py

```diff
@@ -0,0 +1,250 @@
+#!/usr/local/bin/python3
+
+"""
+    Copyright (c) 2023 Deciso B.V.
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+
+    1. Redistributions of source code must retain the above copyright notice,
+     this list of conditions and the following disclaimer.
+
+    2. Redistributions in binary form must reproduce the above copyright
+     notice, this list of conditions and the following disclaimer in the
+     documentation and/or other materials provided with the distribution.
+
+    THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+    INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+    AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import os
+import requests
+import syslog
+import re
+import glob
+import importlib
+import sys
+import fcntl
+import ujson
+import time
+from configparser import ConfigParser
+
+class BaseBlocklistHandler:
+    def __init__(self, config):
+        self.config = config
+        self.cnf = None
+        self.cnf_cache = None
+        self.diffs_added = {}
+        self.diffs_removed = {}
+        self.priority = 0
+
+        self.cur_bl_location = '/var/unbound/data/dnsbl.json'
+
+        self.domain_pattern = re.compile(
+            r'^(([\da-zA-Z_])([_\w-]{,62})\.){,127}(([\da-zA-Z])[_\w-]{,61})'
+            r'?([\da-zA-Z]\.((xn\-\-[a-zA-Z\d]+)|([a-zA-Z\d]{2,})))$'
+        )
+
+        self._load_config()
+
+    def get_config(self):
+        """
+        Get statically defined configuration options.
+        """
+        pass
+
+    def get_blocklist(self):
+        """
+        Overridden by derived classes to produce a formatted blocklist. Returns a dictionary
+        with domains as keys and a dictionary of metadata as values
+        """
+        pass
+
+    def _write_cache(self):
+        """
+        Cache so each derived class can decide whether a download is necessary
+        """
+        if self.cnf:
+            with open('/tmp/unbound-blocklists.conf.cache', 'w') as cache_config:
```

Comment:
@swhite2 I don't think this file is generic as `unbound-blocklists.conf` belongs to `DefaultBlocklistHandler` in order to be able to reuse files, it might even be better to define a "cache" directory and set a ttl on these files to download. This way we could loose part of the complexity of determining if we should download and offer reusability of files when they are downloaded from the same location.

Reply:
> @swhite2 I don't think this file is generic as `unbound-blocklists.conf` belongs to `DefaultBlocklistHandler` 

This is an oversight, it should be `self.config`, not `/tmp/unbound-blocklists.conf.cache`. Each derived class defines its own config file.